File: libraries/typescript/v1/types/criteria.ts
export type Color = string | {
  red: number;
  green: number;
  blue: number;
};

export type QualityStyle = {
  primaryColor: Color;
  secondaryColor: Color;
};

export type SchemaDefinition = Record<string, any>;

export type SchemaDefinitions = {
  parameters?: SchemaDefinition;
  result?: SchemaDefinition;
};

export type Metadata = {
  id: string;
  version: string;
  date: Date;
  name: string;
  description: string;
  documentation?: string;
  locales?: string[];
  defaultLocale?: string;
  schemas?: SchemaDefinitions;
}

export type PdfDocumentationItem = {
  type: 'pdf';
  label: string,
  url: string;
  text: string;
};

export type InlineDocumentationItem = {
  type: 'text';
  label: string;
  text: string;
};

export type LinkDocumentationItem = {
  type: 'link';
  label: string,
  url: string;
  text: string;
};

export type DocumentationItem = PdfDocumentationItem | InlineDocumentationItem | LinkDocumentationItem;

export type CriteriaTree = {
  version: string;
  qualities: Quality[];
  certifications?: string[];
  certificationDefinitions?: CertificationDefinition[];
  result?: any;
};

export type ElementData = Record<string, any> & {
  value?: TaskItemValue;
  text?: string;
};

export type TaskItemData = ElementData & {
  valueReference?: TaskItemValue;
};

type BaseElement<Type extends CriteriaTreeElementType> = {
  type: Type;
  title: string;
  code: string;
  tags?: string[];
  documentation?: DocumentationItem[];
  data?: ElementData;
  sortOrder?: number;
};

export type Quality = BaseElement<'quality'> & {
  style?: QualityStyle;
  items: Criterion[];
}

export type Criterion = BaseElement<'criterion'> & {
  items: TaskGroup[];
};

export type TaskGroup = BaseElement<'task-group'> & {
  category?: string;
  items: Task[];
};

export type Task = BaseElement<'task'> & {
  description?: string;
  items: TaskItem[];
};

export type TaskItem = Omit<BaseElement<'task-item'>, 'title' | 'data'> & {
  definition: TaskItemDefinition;
  description?: string;
  providedData?: Record<string, TaskItemValue>;
  data: TaskItemData;
};

export type AbstractDefinition<Type extends 'select-single' | 'select-multiple' | 'number' | 'boolean'> = {
  type: Type;
}

export type SelectSingleType = AbstractDefinition<'select-single'> & {
  options: PointOption[];
};

export type SelectMultipleType = AbstractDefinition<'select-multiple'> & {
  options: PointOption[];
};

export type NumberType = AbstractDefinition<'number'> & {
  minimum?: number;
  maximum?: number;
  step?: number;
};

export type BooleanType = AbstractDefinition<'boolean'> & {
  labels?: {
    true?: string;
    false?: string;
  }
};

export type PointOption = {
  id?: string;
  text: string;
  intro?: string;
  outro?: string;
  value: string | number | boolean | null;
};

export type TaskItemScalarValue = string | number | boolean | null;
export type TaskItemValue = TaskItemScalarValue | Array<TaskItemScalarValue>;
export type TaskItemDefinition = SelectSingleType | SelectMultipleType | NumberType | BooleanType;

export type CriteriaTreeElement = Quality | Criterion | TaskGroup | Task | TaskItem;
export type CriteriaTreeElementType = 'quality' | 'criterion' | 'task-group' | 'task' | 'task-item';

export type DataMap = {
  version: string;
  elements: Record<string, any>;
  result: any;
  certifications?: string[];
};

export type CertificationDefinitionType = 'number' | 'percentage';

type AbstractCertificationDefinitionRules<Rules> = Rules extends undefined | never
  ? {
    rules?: Rules;
    rulesText?: string;
  }
  : {
    rules: Rules;
    rulesText: string;
  };

type AbstractCertificationDefinition<Type extends CertificationDefinitionType, Rules> = AbstractCertificationDefinitionRules<Rules> & {
  code: string;
  type: Type;
  icon?: string;
  name: string;
  description?: string;
};

export type NumberBasedCertificationDefinitionRules = {
  minimum?: number;
  exclusiveMinimum?: number;
  maximum?: number;
  exclusiveMaximum?: number;
};

export type PercentageBasedCertificationDefinitionRules = NumberBasedCertificationDefinitionRules;

export type NumberBasedCertificationDefinition = AbstractCertificationDefinition<'number', NumberBasedCertificationDefinitionRules>;
export type PercentageBasedCertificationDefinition = AbstractCertificationDefinition<'percentage', PercentageBasedCertificationDefinitionRules>;

export type CertificationDefinition = NumberBasedCertificationDefinition | PercentageBasedCertificationDefinition;

File: libraries/python/v1/criteria.py
import datetime
import typing
from typing import Optional


class Metadata:
    def __init__(self, id: str, version: str, date: datetime.datetime, name: str, description: str, documentation: str):
        self.id = id
        self.version = version
        self.date = date
        self.name = name
        self.description = description
        self.documentation = documentation


class DocumentationItem:
    def __init__(self, type: str, label: str, url: Optional[str], text: str):
        self.type = type
        self.label = label
        self.url = url
        self.text = text


class PdfDocumentationItem(DocumentationItem):
    def __init__(self, label: str, url: str, text: str):
        super().__init__('pdf', label, url, text)


class InlineDocumentationItem(DocumentationItem):
    def __init__(self, label: str, text: str):
        super().__init__('text', label, None, text)


class LinkDocumentationItem(DocumentationItem):
    def __init__(self, label: str, url: str, text: str):
        super().__init__('link', label, url, text)


TaskItemValuePrimitive = typing.Union[str, float, bool, None]
TaskItemValue = typing.Union[TaskItemValuePrimitive, list[TaskItemValuePrimitive]]

DefinitionType = typing.Literal['select-single', 'select-multiple', 'number', 'boolean']


class PointOption:
    def __init__(self, id: Optional[str], label: str, value: TaskItemValuePrimitive, annotation: Optional[str]):
        self.id = id
        self.label = label
        self.value = value
        self.annotation = annotation


class BaseTaskItemDefinition:
    def __init__(self, type: DefinitionType):
        self.type = type


class SelectSingleType(BaseTaskItemDefinition):
    def __init__(self, options: list[PointOption]):
        super().__init__('select-single')
        self.options = options


class SelectMultipleType(BaseTaskItemDefinition):
    def __init__(self, options: list[PointOption]):
        super().__init__('select-multiple')
        self.options = options


class NumberType(BaseTaskItemDefinition):
    def __init__(self, minimum: Optional[float], maximum: Optional[float], step: Optional[float]):
        super().__init__('number')
        self.minimum = minimum
        self.maximum = maximum
        self.step = step


class BooleanType(BaseTaskItemDefinition):
    def __init__(self):
        super().__init__('boolean')


TaskItemDefinition = typing.Union[SelectSingleType, SelectMultipleType, NumberType, BooleanType]
CriteriaTreeElementType = typing.Literal['criterion', 'task-group', 'task', 'task-item']


class BaseElement:
    def __init__(self, type: CriteriaTreeElementType, id: str, label: Optional[str], tags: Optional[list],
                 documentation: Optional[list[DocumentationItem]]):
        self.type = type
        self.id = id
        self.label = label
        self.tags = tags
        self.documentation = documentation


class TaskItem(BaseElement):
    def __init__(self, id: str, label: Optional[str], tags: Optional[list],
                 documentation: Optional[list[DocumentationItem]], description: Optional[str],
                 definition: TaskItemDefinition, provided_data: Optional[dict[str, TaskItemValue]],
                 calculated_data: Optional[dict[str, any]]):
        super().__init__('task-item', id, label, tags, documentation)
        self.description = description
        self.definition = definition
        self.provided_data = provided_data
        self.calculated_data = calculated_data


class Task(BaseElement):
    def __init__(self, id: str, label: Optional[str], tags: Optional[list],
                 documentation: Optional[list[DocumentationItem]], description: Optional[str], items: list[TaskItem]):
        super().__init__('task', id, label, tags, documentation)
        self.description = description
        self.items = items


class TaskGroup(BaseElement):
    def __init__(self, id: str, label: Optional[str], tags: Optional[list],
                 documentation: Optional[list[DocumentationItem]], items: list[Task]):
        super().__init__('task-group', id, label, tags, documentation)
        self.items = items


class Criterion(BaseElement):
    def __init__(self, id: str, label: Optional[str], tags: Optional[list],
                 documentation: Optional[list[DocumentationItem]], quality: str, items: list[TaskGroup]):
        super().__init__('criterion', id, label, tags, documentation)
        self.quality = quality
        self.items = items


class CriteriaTree(list[Criterion]):
    def __init__(self, criteria: list):
        list.__init__(self, criteria)


CriteriaTreeElement = typing.Union[Criterion, TaskGroup, Task, TaskItem]
